GPU utilization in Daytona for LLM fine-tuning and inference

stdthoth opened the pull request:
/claim #8 


## Writer's Checklist

### Writing Structure

- [x] Use short sentences and paragraphs, and include bucket brigades.
- [x] Include more than two descriptive H2 headings to organize content.
- [x] Capitalize headings according to the AP Stylebook
      ([use this tool](https://capitalizemytitle.com/style/AP/))
- [x] Include an introduction with at least two paragraphs before the first H2
      section.
- [x] Use appropriate formatting (bold, italic, underline), notes, quotes,
      TLDRs, and key points.
- [x] Incorporate `code elements` and `Markdown format` where appropriate.
- [x] Ensure at least one visual element per “screen” of the article
      (screenshots, diagrams, tables, graphs, lists, sidenotes, blockquotes).

### Fact-Checking

- [x] Verify all facts and data points included in the article.

### Asset Management

- [x] Save images in the `/assets` folder.
- [x] Follow naming conventions:
      `YYYYMMDD_title_of_the_article_IMG_NAME_NO.png`.
- [x] (Optional) Create a GitHub repo for the code referenced in the article and
      share it.
- [ ] (Optional) Include a link to this Loom video in the PR comments.

### Interlinking

- [x] Use `CTRL+F` to search for relevant keywords in:
  - [Daytona Dotfiles Sitemap](https://www.daytona.io/sitemap-dotfiles.xml)
  - [Daytona Definitions Sitemap](https://www.daytona.io/sitemap-definitions.xml)
- [x] Interlink keywords using the appropriate URLs.

### Glossary/Definitions

- [x] Create new definition in `/defitnitions` folder.

### Review and Edit

- [x] Ensure articles address the needs of the target audience and their search
      intent.
- [x] Read the article out loud to catch any awkward phrasing.
- [x] Run the draft through [Grammarly](https://grammarly.com) or a similar
      grammar tool.
- [x] Double-check adherence to the style guide and repository guidelines.
- [x] Use the name of the article for the title of the PR.

stdthoth committed:
initial commit

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth committed:
clarify driver installation instructions for Windows and Linux users in GPU utilization guide

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth committed:
Update 20250105_daytona_gpu_utilization.md

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth committed:
Add definitions for GPU and LLM in new markdown files

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth committed:
Update 20250105_daytona_gpu_utilization.md

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth committed:
edit dockerfile section

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth commented:
Happy new year @mojafa please this PR is ready for review

stdthoth committed:
Update 20250105_daytona_gpu_utilization.md

Signed-off-by: stdthoth <ichthoth@gmail.com>

mojafa commented:
@stdthoth apologies, just to let you know there's another PR that also tackles the same issue. I'm reviewing all of them and will assign the one with the best quality.

stdthoth commented:
@mojafa no problem.. take your time and I am available for any clarifications

stdthoth commented:
Any updates concerning the review @mojafa

mojafa commented:
@stdthoth Just to let you know, someone else was assigned this task but had to close their PR as he had attempted more than one issue. I'll also check for code similarity.

I'm reviewing your PR now, and if it's sufficient, I'll assign this to you.

mojafa commented:
@stdthoth please link keywords in your intro paragrpahds. it's very plain and some terms like GPU should be linked as a keyword.

mojafa commented:
Also, your TL/DR section is not supposed to be there. 
The intro section (anything before TL/DR) is too long. Follow the template.

I'll test the code, but please ensure that your article follows the template.


Equally, this is an article, not a guide, according to the issue: https://github.com/daytonaio/content/issues/8

stdthoth committed:
Add definitions for LLM, tokenization, GPU utilization, CUDA, GPU driver, and LLM inference and added as keyword in article

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth committed:
references

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth commented:
> Also, your TL/DR section is not supposed to be there. The intro section (anything before TL/DR) is too long. Follow the template.
> 
> I'll test the code, but please ensure that your article follows the template.
> 
> Equally, this is an article, not a guide, according to the issue: #8

- Reduced the intro section and made it more relevant to the subject
- Added new definitions and linked to the respective keyword
- moved article to the article folder
- added references

mojafa commented:
@stdthoth allow me to re-review this over the next 2 days

stdthoth commented:
@mojafa okay thanks 👍🏽

stdthoth commented:
@mojafa any updates ??

mojafa commented:
@stdthoth reviewing now

mojafa commented:
@stdthoth I've assigned this isseu to you. 

However, I'll need more time as i need to provision a GPU based machine to run nvidia.

In the meantime, please share your screen recording of your solution. 

The content of the article lokos good.

mojafa commented:
@stdthoth pleaae message me on slack @jafa 

need your time on provisioning some VM

stdthoth commented:
@mojafa could you send a link to my email.. the link seems to have expired

stdthoth commented:
@mojafa sent

mojafa commented:
@stdthoth I managed to spin up a GPU enabled VM but I'm facing some issues.


I feel like your article jumps directly into the technical steps without providing some essential context upfront, such as system architecture requirements and prerequisites.

My suggestions:
1. Add more context in the introductory section to introduce the reader to the necessary system specifications and prerequisites in a clearer way. This could include a brief overview of the hardware, software, and environment compatibility.
2. Talk about the **System Architecture**: Before diving into the installation steps, briefly mention the supported architectures (e.g., x86_64 vs ARM) and CUDA compatibility, as this can impact the ability to run GPU-based tasks.
3. **Clarify Prerequisites**: In the "Prerequisites" section, list the minimum requirements such as:
   - **System Architecture**: AMD64 (x86_64) or ARM (e.g., NVIDIA Jetson devices).
   - **Operating System**: Ubuntu or other Linux distributions (with specific versions).
   - **Hardware Requirements**: NVIDIA GPU with CUDA support (e.g., RTX series or higher).
   - **Software Requirements**: Docker, WSL2 for Windows users, IDEs (VS Code/JetBrains), CUDA drivers.
4. **Verify GPU Compatibility**: Make sure to highlight the importance of checking the GPU's architecture and whether it's supported by CUDA, as this directly affects the ability to run GPU-enabled containers.
5. **Diagram or Visualization**: Consider including a visual representation of how each piece of the environment fits together, such as:
   - **High-level architecture of GPU utilization**.
   - **System requirements** (OS version, GPU, etc.).
6. **Define Key Terms Early**: Some terms like "CUDA-enabled GPU," "Daytona," "LLM," and "Fine-Tuning" are defined later, but it might help to give brief explanations earlier in the article. A quick glossary at the start could help the reader grasp the terminology before diving into technical details.
7. **Guide for Different OS Users**: Since you're dealing with different environments (Linux and Windows), you could add a section like "Special Considerations for Linux vs. Windows" early on to guide users based on their system.
8. **Pre-installation Verification**: You can have a section to verify that the prerequisites are met before proceeding, such as checking if Docker, WSL2, NVIDIA drivers, and CUDA are installed and functioning.
9. **Final Walkthrough/Conclusion**: Wrap up with a brief summary or checklist that recaps the setup process and highlights the steps that users should verify to ensure everything is set up correctly.

---

### Revised Example Intro:

#### Prerequisites and System Architecture

Before starting the setup process for GPU-based LLM fine-tuning and inference with Daytona, ensure your system meets the following requirements:

**Hardware**:
- **CPU**: x86_64 (AMD64) or ARM architecture (e.g., NVIDIA Jetson devices).
- **GPU**: CUDA-compatible NVIDIA GPU (e.g., RTX 20xx, 30xx, or A100).
- **Memory**: Minimum of 8GB RAM recommended, 16GB or more for better performance.

**Software**:
- **Operating System**: Ubuntu 18.04+ or other Linux distributions (Windows users will use WSL2).
- **Docker**: Must be installed and configured for running containers.
- **NVIDIA Drivers**: Ensure the correct GPU drivers for your hardware are installed. You can verify this by running `nvidia-smi`.

**Environment**:
- **CUDA**: Install the correct version of the CUDA toolkit matching your GPU drivers. Ensure compatibility between Docker images and the driver version.
- **WSL2**: For Windows users, ensure that WSL2 is properly configured to allow GPU access within Linux environments.

#### TL;DR

This guide will walk you through setting up a Daytona-powered environment for LLM fine-tuning and inference using your GPU. It assumes you have a CUDA-compatible GPU and a Linux-based system (Windows users need to enable WSL2). You’ll use Docker containers, Daytona, and VS Code to streamline the process and leverage your GPU for faster model training and inference.

---

This way, the reader gets a full picture right from the start!

mojafa commented:
@stdthoth running into this error
<img width="1118" alt="Screenshot 2025-02-03 at 11 18 33" src="https://github.com/user-attachments/assets/e6081cdc-ae50-40b5-b2d7-c2201ea83797" />

mojafa commented:
had to use: `docker run --gpus all nvidia/cuda:12.0.1-base-ubuntu22.04 nvidia-smi`


<img width="919" alt="Screenshot 2025-02-03 at 11 21 26" src="https://github.com/user-attachments/assets/5178c86f-df5b-4045-9262-caff545ca4fe" />

mojafa commented:
@stdthoth please hvae a look at your dev configs, running `daytona create` fails to create the workspace. 

Please test lo

<img width="1167" alt="Screenshot 2025-02-03 at 11 28 56" src="https://github.com/user-attachments/assets/ffbfb8c8-e415-4566-90d8-618a0c37320b" />

stdthoth commented:
@mojafa already making edits

stdthoth committed:
considerations

Signed-off-by: stdthoth <ichthoth@gmail.com>

stdthoth commented:
@mojafa I have made all the necessary changes you requested 👌🏽

stdthoth commented:
@mojafa concerning the workspace failing to start. I made a mistake with the name of the .txt file with which the llm uses for fine-tuning.. please change `train.txt` to `data.txt` in the Dockerfile and also in the python file for inference. Apologies for the error

stdthoth committed:
changed train.txt to data.txt

Signed-off-by: stdthoth <ichthoth@gmail.com>

nkkko commented:
I have tried reading the article, but it is not the expected quality. It is hard to read, and the introduction misses the point. Please read the connected issue and think about how to rewrite the article more fluently.

utilization
/ˌjuːtɪlʌɪˈzeɪʃn/
noun
the action of making practical and effective use of something.

...and not "a percentage of a [GPU's]" - this is nonsense....

I appreciate the hard work, but this needs to be properly written, have a narrative and a story arc. And in the end make sense. @stdthoth

stdthoth commented:
@nkko I am currently working on this, can you please reopen.. I'll have something by tomorrow
